Description: Regulate dual-stacking of listener.
 When the remote host is targeted using IPv6, and
 'net.ipv6.bindv6only=0' holds, then the local
 listening socket will accept both address families.
 As this intrudes on the available socket name space,
 an additional test can be used to reset the IPv6
 socket to be single stacked.
Author: Mats Erik Andersson <debian@gisladisker.se>
Forwarded: yes
Last-Updated: 2011-07-20

diff -Naur netsed-1.00b.debian/netsed.c netsed-1.00b/netsed.c
--- netsed-1.00b.debian/netsed.c
+++ netsed-1.00b/netsed.c
@@ -395,6 +395,11 @@
       continue;
     setsockopt(lsock, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
     //fcntl(lsock,F_SETFL,O_NONBLOCK);
+    /* Make our best to decide on dual-stacked listener. */
+    one = (af == 0) ? 0 /* AF_UNSPEC given */ : 1; /* Preconditioned addr */
+    if (res->ai_family == AF_INET6)
+      if (setsockopt(lsock, IPPROTO_IPV6, IPV6_V6ONLY, &one, sizeof(one)))
+        printf("    Failed to unset IPV6_V6ONLY: %s.\n", strerror(errno));
     if (bind(lsock, res->ai_addr, res->ai_addrlen) < 0) {
       ERR("bind(): %s", strerror(errno));
       close(lsock);
